File: packages/zynth-sensors/android/ZynthSensors/src/main/java/dev/zynth/sensors/ZynthSensors.kt
package dev.zynth.sensors

import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import com.zynth.kit.runtime.ZynthRuntime

object ZynthSensors {
    private const val TAG = "ZynthSensors"
    private var moduleInstance: SensorsModule? = null

    @JvmStatic
    fun initialize(activity: ComponentActivity, runtime: ZynthRuntime) {
        if (moduleInstance != null) {
            Log.w(TAG, "Module already initialized")
            return
        }

        val module = SensorsModule(runtime, activity)
        val permissionLauncher: ActivityResultLauncher<String> =
            activity.registerForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
                module.onPermissionResult(granted)
            }

        module.permissionLauncher = permissionLauncher
        moduleInstance = module

        runtime.installModules(listOf(module))
        Log.d(TAG, "Module initialized")
    }

    @JvmStatic
    fun cleanup() {
        moduleInstance?.invalidate()
        moduleInstance = null
    }
}

File: packages/zynth-sensors/ios/ZynthSensors.swift
import Foundation
import ZynthKit

@objc(ZynthSensors)
public class ZynthSensors: NSObject {
  private static var moduleInstance: SensorsModule?

  @objc public static func initialize(with runtime: ZynthRuntime) {
    guard moduleInstance == nil else {
      NSLog("[ZynthSensors] Module already initialized")
      return
    }

    let module = SensorsModule(runtime: runtime)
    runtime.installModules([module])
    moduleInstance = module
    NSLog("[ZynthSensors] Module initialized")
  }

  @objc public static func cleanup() {
    moduleInstance?.invalidate()
    moduleInstance = nil
  }
}
